### PR TITLE
Cache recipient hash lookup for custom domains

### LINKED ISSUE #2863: Cache recipient hash lookup for custom domains
## Problem Statement

Custom domain incoming secrets recompute SHA256 hashes for all configured recipients on every lookup request. With up to 20 recipients per domain, this means 20 SHA256 operations per API call. In contrast, canonical domain recipients use a frozen pre-built hash map populated at boot time with O(1) lookup.

## Early Diagnosis

The asymmetry appears intentional — canonical recipients are static (YAML config), while custom domain recipients can change at runtime. However, the `IncomingSecretsConfig` instance is already loaded from Redis per-request; caching the computed hash map on that instance (or memoizing at the CustomDomain level with cache invalidation on update) would align performance characteristics.

Likely not a problem at current scale, but worth aligning the design before custom domain adoption grows.

## Affected Areas

- [ ] `lib/onetime/models/custom_domain/incoming_secrets_config.rb` — `lookup_incoming_recipient` and `public_incoming_recipients` recompute hashes
- [ ] `lib/onetime/incoming/recipient_resolver.rb` — calls config methods, could cache result
- [ ] `apps/api/v3/initializers/setup_incoming_recipients.rb` — reference implementation for boot-time caching

## Parent

Part of #2525

### PR DESCRIPTION
**Pull request by delano**
## Summary

Custom domain incoming secrets were recomputing SHA256 hashes for all configured recipients on every API call (up to 20 operations per request). This adds instance-level caching to align with the canonical domain pattern that uses frozen pre-built hash maps for O(1) lookup.

- Add `cached_incoming_recipient_lookup` and `cached_public_incoming_recipients` methods to `CustomDomain`
- Cache keyed by site_secret hash to support multiple secrets (key rotation)
- Automatic cache invalidation when `update_incoming_secrets_config` is called
- Returns frozen objects for thread safety

## Test plan

- [x] New tryout test suite with 28 test cases covering caching behavior, invalidation, edge cases
- [x] Existing recipient resolver tests pass (65 tests)
- [ ] Manual: verify `/api/incoming/config` returns correct public recipients
- [ ] Manual: verify `/api/incoming/validate` resolves recipients correctly after config update

Closes #2863

### CODE REVIEW

File: lib/onetime/models/custom_domain.rb [hunk 1]
```diff
@@ -458,14 +458,48 @@ def incoming_secrets_config
...
+    #
+    # Keyed by site_secret hash to support multiple secrets (e.g., key rotation).
+    # Returns frozen hash for thread safety and to match canonical domain pattern.
+    #
+    # @param site_secret [String] Site secret used as hash salt
+    # @return [Hash<String, String>] Frozen hash mapping recipient hashes to emails
+    def cached_incoming_recipient_lookup(site_secret)
+      return {}.freeze if site_secret.nil? || site_secret.to_s.strip.empty?
+
+      cache_key                           = Digest::SHA256.hexdigest(site_secret)
```
[the remark this answers is not in the document]
Comment (delano) [reply lib/onetime/models/custom_domain.rb:472]: This alignment was applied by RuboCop's `Layout/ExtraSpacing` auto-correction during pre-commit hooks. The project follows RuboCop's style configuration — dismissing this suggestion to maintain consistency with the linter.

File: lib/onetime/models/custom_domain.rb [hunk 2]
```diff
@@ -458,14 +458,48 @@ def incoming_secrets_config
...
     # @return [void]
     def update_incoming_secrets_config(config)
-      @incoming_secrets_config = nil # clear cache
-      self.incoming_secrets    = config.to_json
-      self.updated             = OT.now.to_i
+      @incoming_secrets_config           = nil # clear config cache
+      @incoming_recipient_lookup_cache   = nil # clear lookup cache
+      @public_incoming_recipients_cache  = nil # clear public recipients cache
+      self.incoming_secrets              = config.to_json
+      self.updated                       = OT.now.to_i
```
[the remark this answers is not in the document]
Comment (delano) [reply lib/onetime/models/custom_domain.rb:502]: Same as above — this alignment style is RuboCop's auto-correction. The project's linter configuration takes precedence over manual style preferences.
